tools/wptrunner/wptrunner/executors/base.py

```diff
@@ -515,45 +520,63 @@ def run_test(self, test):
                 if success is False:
-                    return {"status": data[0], "message": data[1]}
+                    test_results = {"status": data[0], "message": data[1]}
+                    if self.reftest_screenshot == "always" and log_data:
+                        test_results["extra"] = {"reftest_screenshots": log_data}
+                    return test_results
 
                 hashes[i], screenshots[i] = data
                 urls[i] = node.url
 
             is_pass, page_idx = self.check_pass(hashes, screenshots, urls, relation, fuzzy)
+
             if is_pass:
+                if self.reftest_screenshot == "always":
+                    log_data.extend(
+                        [{"url": urls[0], "screenshot": screenshots[0][page_idx],
+                          "hash": hashes[0][page_idx]},
+                         relation,
+                         {"url": urls[1], "screenshot": screenshots[1][page_idx],
+                          "hash": hashes[1][page_idx]}])
+
                 fuzzy = self.get_fuzzy(test, nodes, relation)
                 if nodes[1].references:
                     stack.extend(list(((nodes[1], item[0]), item[1])
                                       for item in reversed(nodes[1].references)))
                 else:
+                    test_result = {"status": "PASS", "message": None}
+                    if self.reftest_screenshot == "always":
+                        test_result["extra"] = {"reftest_screenshots": log_data}
                     # We passed
-                    return {"status": "PASS", "message": None}
+                    return test_result
 
         # We failed, so construct a failure message
 
-        if page_idx is None:
-            # default to outputting the last page
-            page_idx = -1
         for i, (node, screenshot) in enumerate(zip(nodes, screenshots)):
             if screenshot is None:
                 success, screenshot = self.retake_screenshot(node, viewport_size, dpi, page_ranges)
                 if success:
                     screenshots[i] = screenshot
 
-        log_data = [
-            {"url": nodes[0].url,
-             "screenshot": screenshots[0][page_idx],
-             "hash": hashes[0][page_idx]},
-            relation,
-            {"url": nodes[1].url,
-             "screenshot": screenshots[1][page_idx],
-             "hash": hashes[1][page_idx]},
-        ]
-
-        return {"status": "FAIL",
-                "message": "\n".join(self.message),
-                "extra": {"reftest_screenshots": log_data}}
+        test_result =  {"status": "FAIL",
+                        "message": "\n".join(self.message)}
+        if self.reftest_screenshot == "always":
+            test_result["extra"] = {"reftest_screenshots": log_data}
+        elif (not self.reftest_screenshot or
+              self.reftest_screenshot == "fail" or
+              (self.reftest_screenshot == "unexpected" and
+               test.expected() == 'PASS')):
```

Comment:
We should write the comparison with `expected()` like `test.expected() != "FAIL"` rather than assuming that only an expected `PASS` could require a screenshot if the result is `FAIL`.

Reply:
Done